File: c/Cpp/CLexerBase.cpp
#include "CLexerBase.h"

#include <algorithm>
#include <array>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>

CLexerBase::CLexerBase(antlr4::CharStream *input)
    : CLexerBase(input, runGccAndMakeStream(input)) {}

CLexerBase::CLexerBase(antlr4::CharStream *original,
                       std::unique_ptr<antlr4::ANTLRInputStream> preprocessed)
    : antlr4::Lexer(preprocessed ? preprocessed.get() : original)
    , preprocessedStream_(std::move(preprocessed)) {}

//  Static configuration

void CLexerBase::setArgs(const std::vector<std::string>& args) {
    args_ = args;
}

//  runGccAndMakeStream()

std::unique_ptr<antlr4::ANTLRInputStream>
CLexerBase::runGccAndMakeStream(antlr4::CharStream *input)
{
    bool vsc   = hasArg("--vsc");
    bool gcc   = hasArg("--gcc");
    bool clang = hasArg("--clang");
    bool nopp  = hasArg("--nopp");

    // Default to gcc
    if (!(vsc || gcc || clang)) {
        gcc = true;
    }

    auto ppOptions = extractPreprocessorOptions();

    // ---- source name / text -----------------------------------------------
    std::string sourceName = input->getSourceName();
    if (sourceName.empty() || sourceName.size() < 2 ||
        sourceName.substr(sourceName.size() - 2) != ".c")
    {
        sourceName = "stdin.c";
    }
    std::string outputName = sourceName + ".p";
    std::string inputText;
    auto inputSize = input->size();
    if (inputSize > 0) {
        inputText = input->getText(
            antlr4::misc::Interval(0, static_cast<ssize_t>(inputSize) - 1));
    }

    // ---- --nopp -----------------------------------------------------------
    if (nopp) {
        { std::ofstream out(outputName); if (out) out << inputText; }
        auto stream = std::make_unique<antlr4::ANTLRInputStream>(inputText);
        stream->name = sourceName;
        return stream;
    }

    // Persist stdin text so external preprocessor can read it.
    if (sourceName == "stdin.c") {
        std::ofstream out(sourceName);
        if (out) out << inputText;
    }

    // ---- gcc / clang (external) -------------------------------------------
    if (gcc || clang) {
        std::string cmd = gcc ? "gcc" : "clang";

        std::ostringstream cmdline;
        cmdline << cmd << " -std=c2x -E -C";
        for (auto& opt : ppOptions) {
            cmdline << " " << opt;
        }
        cmdline << " " << sourceName << " 2>/dev/null";

        FILE *pipe = popen(cmdline.str().c_str(), "r");
        if (!pipe) {
            throw std::runtime_error(
                std::string("Failed to run ") + cmd + " preprocessor");
        }

        std::ostringstream captured;
        std::array<char, 4096> buf{};
        while (std::fgets(buf.data(), static_cast<int>(buf.size()), pipe)) {
            captured << buf.data();
        }
        int rc = pclose(pipe);
        (void)rc;

        std::string preprocessed = captured.str();

        { std::ofstream out(outputName); if (out) out << preprocessed; }

        auto stream = std::make_unique<antlr4::ANTLRInputStream>(preprocessed);
        stream->name = sourceName;
        return stream;
    }

    // Note: vsc (Visual Studio cl.exe) is Windows-specific; not ported.
    throw std::runtime_error("No preprocessor specified.");
}

//  Argument helpers

bool CLexerBase::hasArg(const std::string& arg) {
    std::string lower_arg = arg;
    std::transform(lower_arg.begin(), lower_arg.end(),
                   lower_arg.begin(), ::tolower);
    for (auto& a : args_) {
        std::string lower_a = a;
        std::transform(lower_a.begin(), lower_a.end(),
                       lower_a.begin(), ::tolower);
        if (lower_a.find(lower_arg) != std::string::npos) {
            return true;
        }
    }
    return false;
}

std::vector<std::string> CLexerBase::extractPreprocessorOptions() {
    std::vector<std::string> options;
    for (auto& arg : args_) {
        if (arg.starts_with("--D")) {
            options.push_back("-D" + arg.substr(3));
        } else if (arg.starts_with("--I")) {
            options.push_back("-I" + arg.substr(3));
        }
    }
    return options;
}

File: c/Go/CLexerBase.go
package parser

import (
	"os"
	"os/exec"
	"runtime"
	"strings"

	"github.com/antlr4-go/antlr/v4"
)

type CLexerBase struct {
	*antlr.BaseLexer
	preprocessed bool
}

// NextToken overrides BaseLexer.NextToken to run the C preprocessor once
// before any token is consumed, mirroring what the C# and Java CLexerBase
// constructors do with RunGccAndMakeStream / runGccAndMakeStream.
func (l *CLexerBase) NextToken() antlr.Token {
	if !l.preprocessed {
		l.preprocessed = true
		newInput := runGccAndMakeStream(l.GetInputStream())
		l.SetInputStream(newInput)
	}
	return l.BaseLexer.NextToken()
}

func runGccAndMakeStream(input antlr.CharStream) antlr.CharStream {
	isWindows := runtime.GOOS == "windows"
	args := os.Args

	vsc := hasArg(args, "--vsc")
	gcc := hasArg(args, "--gcc")
	clang := hasArg(args, "--clang")
	nopp := hasArg(args, "--nopp")

	if !(vsc || gcc || clang) {
		gcc = true
	}

	ppOptions := extractPreprocessorOptions(args)

	sourceName := input.GetSourceName()
	if sourceName == "" || sourceName == "<unknown source>" || !strings.HasSuffix(sourceName, ".c") {
		sourceName = "stdin.c"
	}
	outputName := sourceName + ".p"
	inputText := input.GetText(0, input.Size()-1)

	if nopp {
		_ = os.WriteFile(outputName, []byte(inputText), 0644)
		return antlr.NewInputStream(inputText)
	}

	if sourceName == "stdin.c" {
		_ = os.WriteFile(sourceName, []byte(inputText), 0644)
	}

	if gcc {
		gccCmd := "gcc"
		if isWindows {
			gccCmd = "gcc.exe"
		}
		cmdArgs := []string{"-std=c2x", "-E", "-C"}
		cmdArgs = append(cmdArgs, ppOptions...)
		cmdArgs = append(cmdArgs, sourceName)
		output, err := exec.Command(gccCmd, cmdArgs...).Output()
		if err == nil && len(output) > 0 {
			_ = os.WriteFile(outputName, output, 0644)
			return antlr.NewInputStream(string(output))
		}
		// Fall back to original input if preprocessing fails.
		return antlr.NewInputStream(inputText)
	}

	if clang {
		clangCmd := "clang"
		if isWindows {
			clangCmd = "clang.exe"
		}
		cmdArgs := []string{"-std=c2x", "-E", "-C"}
		cmdArgs = append(cmdArgs, ppOptions...)
		cmdArgs = append(cmdArgs, sourceName)
		output, err := exec.Command(clangCmd, cmdArgs...).Output()
		if err == nil && len(output) > 0 {
			_ = os.WriteFile(outputName, output, 0644)
			return antlr.NewInputStream(string(output))
		}
		// Fall back to original input if preprocessing fails.
		return antlr.NewInputStream(inputText)
	}

	panic("no preprocessor specified")
}

func hasArg(args []string, arg string) bool {
	for _, a := range args {
		if strings.Contains(strings.ToLower(a), arg) {
			return true
		}
	}
	return false
}

func extractPreprocessorOptions(args []string) []string {
	var options []string
	for _, arg := range args {
		if strings.HasPrefix(arg, "--D") {
			options = append(options, "-D"+arg[3:])
		} else if strings.HasPrefix(arg, "--I") {
			options = append(options, "-I"+arg[3:])
		}
	}
	return options
}
